[JSC] Use non-temporal store-pair for stack sanitization

Constellation opened the pull request:
#### cc54113581bd12d60f484c2b349054672fc6fc65
<pre>
[JSC] Use non-temporal store-pair for stack sanitization
<a href="https://bugs.webkit.org/show_bug.cgi?id=244443">https://bugs.webkit.org/show_bug.cgi?id=244443</a>

Reviewed by Mark Lam.

This patch adds ARM64 specific sanitizeStackForVMImpl implementation.
We use stnp to zero-fill unused stack region. It fills 16-byte per one opcode.
And we use non-temporal store-pair (stnp instead of stp) here.
Non-temporal store pair (it only exists for stp and ldp) can fill memory while
we do not keep the new content in the CPU cache. This is
useful if we would like to clear memory which will not be read back soon.
In this case, this is unused stack region, and whenever we use
stack, we fill it with a new value, thus, we do not care about the written
content (in this case, zero). This avoids polluting CPU cache.

0.32% better on Speedometer2 with 80% confidence, but individual test gets a solid
improvement (React-TodoMVC gets 0.58% improvement with 95% confidence).

* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::addressOfLastStackTop): Deleted.

Canonical link: <a href="https://commits.webkit.org/253967@main">https://commits.webkit.org/253967@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bb151e09ae9777944143833bd62be161d8c0f239

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31742 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/18409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96859 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/150621 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30100 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79752 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/91610 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93273 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/24276 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/79287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/18409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67162 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/79468 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27798 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/18409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73174 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27762 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/18409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26037 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2792 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29455 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76005 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29360 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/18409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16855 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1776,11 +1776,29 @@ if not (C_LOOP or C_LOOP_WIN)
+            assert(macro (ok)  btpz a0, (PtrSize * 2 - 1), ok end)
+            btpz address, (PtrSize * 2 - 1), .zeroFillLoop
+            # If it is not aligned, then store pointer-size and increment.
+            emit "str xzr, [x1], #8" # address is a1, thus x1
```
str does not have non-temporal store version.

Constellation commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1776,11 +1776,29 @@ if not (C_LOOP or C_LOOP_WIN)
+        .zeroFillLoop:
+            # Use non-temporal store-pair (stnp) since these stack values are meaningless to the execution.
+            # Avoid poluting CPU cache by using stnp.
+            emit "stnp xzr, xzr, [x1]" # address is a1, thus x1
```
stnp does not have post-increments.

MenloDorian commented on Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
```diff
@@ -1776,11 +1776,29 @@ if not (C_LOOP or C_LOOP_WIN)
+            assert(macro (ok)  btpz address, (PtrSize * 2 - 1), ok end)
+        .zeroFillLoop:
+            # Use non-temporal store-pair (stnp) since these stack values are meaningless to the execution.
+            # Avoid poluting CPU cache by using stnp.
```
/poluting/polluting/

Constellation replied:
Fixed.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Use non-temporal store-pair for stack sanitization
https://bugs.webkit.org/show_bug.cgi?id=244443

Reviewed by Mark Lam.

This patch adds ARM64 specific sanitizeStackForVMImpl implementation.
We use stnp to zero-fill unused stack region. It fills 16-byte per one opcode.
And we use non-temporal store-pair (stnp instead of stp) here.
Non-temporal store pair (it only exists for stp and ldp) can fill memory while
we do not keep the new content in the CPU cache. This is
useful if we would like to clear memory which will not be read back soon.
In this case, this is unused stack region, and whenever we use
stack, we fill it with a new value, thus, we do not care about the written
content (in this case, zero). This avoids polluting CPU cache.

0.32% better on Speedometer2 with 80% confidence, but individual test gets a solid
improvement (React-TodoMVC gets 0.58% improvement with 95% confidence).

* Source/JavaScriptCore/llint/LowLevelInterpreter.asm:
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::addressOfLastStackTop): Deleted.

Canonical link: https://commits.webkit.org/253967@main
